NULL loaded on NOT NULL using load csv

separated this issue from incorrect size reporting that was closed https://github.com/cockroachdb/cockroach/issues/17355#issuecomment-326130454

Will test again when a new alpha is available
[logs.tar.gz](https://github.com/cockroachdb/cockroach/files/1267650/logs.tar.gz)
 as alpha 0817 regressed and cannot load the files

**BUG REPORT**

1. Please supply the header (i.e. the first few lines) of your most recent

$ cockroach version
Build Tag:    v1.1-alpha.20170810
Build Time:   2017/08/10 15:13:13
Distribution: CCL
Platform:     darwin amd64
Go Version:   go1.8.3
C Compiler:   4.2.1 Compatible Clang 3.8.0 (tags/RELEASE_380/final)
Build SHA-1:  c3fca980b5a31384599aba0091cd6a0200e12797
Build Type:   release

2. Please describe the issue you observed:

- What did you do?

  ALTER TABLE lineitem ADD COLUMN interest DECIMAL NOT NULL DEFAULT (DECIMAL '1.3');

- What did you expect to see?

- What did you see instead?

 ALTER TABLE lineitem ADD COLUMN interest DECIMAL NOT NULL DEFAULT (DECIMAL '1.3');
driver: bad connection
connection lost; opening new connection: all session settings will be lost
error retrieving the transaction status: dial tcp [::1]:26257: getsockopt: connection refused
connection lost; opening new connection: all session settings will be lost
error retrieving the database name: dial tcp [::1]:26257: getsockopt: connection refused

```
W170831 11:32:53.636484 280 storage/node_liveness.go:379  [n1,hb] slow heartbeat took 1.8s
panic: Non-nullable column with no value!

goroutine 14541 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).finalizeRow(0xc4203f2f10)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:588 +0x259
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRow(0xc4203f2f10, 0x8841a08, 0xc42a7be800, 0xc420c06b00, 0x40133c8, 0x700, 0x56396a0, 0x1, 0xc420e02e00)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:559 +0x13d
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRowDecoded(0xc4203f2f10, 0x8841a08, 0xc42a7be800, 0xc4657ac600, 0x1, 0x1, 0x1, 0x1, 0xc8)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:569 +0x4c
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*columnBackfiller).runChunk.func1(0x8841a08, 0xc42a7be800, 0xc46abf0b00, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/columnbackfiller.go:209 +0x8b0
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn.func1(0x8841a08, 0xc42a7be800, 0xc46abf0b00, 0xc4211daf48, 0x54fb480, 0xc4211daf80)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:517 +0x43
github.com/cockroachdb/cockroach/pkg/internal/client.(*Txn).Exec(0xc46abf0b00, 0x8841a08, 0xc42a7be800, 0xc42a010101, 0xc4211daf90, 0xc421584400, 0x6b910d0)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/txn.go:749 +0xb6
github.com/cockroachdb/cockroach/pkg/internal/client.(*DB).Txn(0xc42041c840, 0x8841a08, 0xc42a7be800, 0xc452479180, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/pkg/internal/client/db.go:518 +0x142
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*columnBackfiller).runChunk(0xc4203f2c00, 0x88204f8, 0xc44e9c3f50, 0xc4214f3000, 0x1, 0x1, 0xc45bf00c90, 0x9, 0x10, 0xc4211711d0, ...)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/columnbackfiller.go:246 +0x164
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).mainLoop(0xc4203f2c00, 0x88204f8, 0xc44e9c3f50, 0x569446c, 0x8)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:137 +0x5b9
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*backfiller).Run(0xc4203f2c00, 0x88204f8, 0xc44e9c3f50, 0xc42d504430)
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/backfiller.go:89 +0x224
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:366 +0x332

```

I was able to reproduce this. I made a 3 node cluster on my local machine and ran the restore, which first produced the error

```
pq: result is ambiguous (job lease expired)
```

But when run again worked:

```
root@:25001/csv> RESTORE csv.* FROM 'nodelocal:///home/mjibson/load-sst/lineitem.sst';
+--------------------+-----------+--------------------+---------+---------------+----------------+-----------+
|       job_id       |  status   | fraction_completed |  rows   | index_entries | system_records |   bytes   |
+--------------------+-----------+--------------------+---------+---------------+----------------+-----------+
| 277251030137274369 | succeeded |                  1 | 6001225 |      48009720 |              0 | 829069818 |
+--------------------+-----------+--------------------+---------+---------------+----------------+-----------+
(1 row)

Time: 1m14.252081767s

root@:25001/csv> select sum(l_extendedprice) from lineitem;
driver: bad connection
connection lost; opening new connection: all session settings will be lost
error retrieving the transaction status: dial tcp [::1]:25001: getsockopt: connection refused
connection lost; opening new connection: all session settings will be lost
error retrieving the database name: dial tcp [::1]:25001: getsockopt: connection refused
```

Node 1 printed:

```
panic: Non-nullable column "lineitem:l_extendedprice" with no value!

goroutine 12028 [running]:
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).finalizeRow(0xc424b467c8)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:588 +0x318
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.(*RowFetcher).NextRow(0xc424b467c8, 0x2995fa0, 0xc4211ee780, 0xc424674e00, 0x10, 0x10, 0x0, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/rowfetcher.go:559 +0x13f
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*tableReader).Run(0xc424b46000, 0x2995fa0, 0xc4211ee780, 0xc42530bd98)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/tablereader.go:206 +0x3ad
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).Start
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/flow.go:394 +0x3d7
```

I restarted node 1 and reran the query and it succeeded:

```
root@:25001 ?> 
connection lost; opening new connection: all session settings will be lost
root@:25001/csv> 
root@:25001/csv> SELECT sum(l_extendedprice) FROM lineitem;
+----------------------+
| sum(l_extendedprice) |
+----------------------+
|      229577310901.20 |
+----------------------+
(1 row)

Time: 12.905347788s
```

When trying with the dataset from azure at backup-test/benchmarks/tpch/scalefactor-1, the following error happens (and upon retry, continues to happen; also remember that this error occurred once above, too):

```
root@:25001/> restore tpch.lineitem from 'nodelocal:///home/mjibson/az-tpch';
pq: result is ambiguous (job lease expired)
```

And the logs had

```
W170906 04:19:51.736931 167 sql/jobs/registry.go:329  job 277268164244996097: canceling due to liveness failure
W170906 04:19:51.737552 13597 storage/replica.go:2669  [n1,s1,r168/1:/Table/53/5/9{155/22…-613/22…}] context cancellation after 6.9s of attempting command AddSSTable [/Table/53/5/9155/221449809128980481/0,/Table/53/5/9613/221412575599132673/0/NULL)
W170906 04:19:51.737957 13596 ccl/storageccl/import.go:195  [n1,s1,r168/1:/Table/53/5/9{155/22…-613/22…}] addsstable [/Table/53/5/9155/221449809128980481/0,/Table/53/5/9613/221412575599132673/0/NULL) attempt 0 failed: result is ambiguous (context canceled)
E170906 04:19:51.738775 12901 sql/jobs/registry.go:315  job 277268164244996097: ignoring FinishedWith error: result is ambiguous (job lease expired)
```

Have not yet done testing with a real cluster, but wanted to document the current findings with a cluster on a single machine.

--- a/pkg/ccl/sqlccl/csv.go
+++ b/pkg/ccl/sqlccl/csv.go
@@ -524,9 +524,23 @@ func writeRocksDB(
 
 	var kv engine.MVCCKeyValue
 	kv.Key.Timestamp.WallTime = walltime
+	// firstKey is always the first key of the span. lastKey, if nil, means the
+	// current SST hasn't yet filled up. Once the SST has filled up, lastKey is
+	// set to the key at which to stop adding KVs. We have to do this because
+	// all column families for a row must be in one SST and the SST may have
+	// filled up with only some of the KVs from the column families being added.
 	var firstKey, lastKey roachpb.Key
 	var count int64
-	for it.Rewind(); ; it.Next() {
+
+	it.Rewind()
+	if ok, err := it.Valid(); err != nil {
+		return 0, err
+	} else if !ok {
+		return 0, errors.New("could not get first key")
+	}
+	firstKey = it.Key()
+
+	for ; ; it.Next() {
 		if ok, err := it.Valid(); err != nil {
 			return 0, err
 		} else if !ok {
@@ -534,34 +548,34 @@ func writeRocksDB(
 		}
 		count++
 
-		// Save the first key for the span.
-		if firstKey == nil {
-			firstKey = it.Key()
-
-			// Ensure the first key doesn't match the last key of the previous SST.
-			if firstKey.Equal(lastKey) {
-				return 0, errors.Errorf("duplicate key: %s", firstKey)
-			}
-		}
-
 		kv.Key.Key = it.UnsafeKey()
 		kv.Value = it.UnsafeValue()
+		if lastKey != nil {
+			if kv.Key.Key.Compare(lastKey) >= 0 {
+				if err := writeSST(firstKey, lastKey); err != nil {
+					return 0, err
+				}
+				firstKey = it.Key()
+				lastKey = nil
 
+				sst, err = engine.MakeRocksDBSstFileWriter()
+				if err != nil {
+					return 0, err
+				}
+				defer sst.Close()
+			}
+		}
 		if err := sst.Add(kv); err != nil {
 			return 0, errors.Wrapf(err, errSSTCreationMaybeDuplicateTemplate, kv.Key.Key)
 		}
-		if sst.DataSize > sstMaxSize {
-			if err := writeSST(firstKey, kv.Key.Key.Next()); err != nil {
-				return 0, err
-			}
-			firstKey = nil
-			lastKey = append([]byte(nil), kv.Key.Key...)
-
-			sst, err = engine.MakeRocksDBSstFileWriter()
+		if sst.DataSize > sstMaxSize && lastKey == nil {
+			// When we would like to split the file, proceed until we aren't in the
+			// middle of a row. Start by finding the next safe split key.
+			lastKey, err = keys.EnsureSafeSplitKey(kv.Key.Key)
 			if err != nil {
 				return 0, err
 			}
-			defer sst.Close()
+			lastKey = lastKey.PrefixEnd()
 		}
 	}
 	if sst.DataSize > 0 {

--- a/pkg/ccl/sqlccl/csv_test.go
+++ b/pkg/ccl/sqlccl/csv_test.go
@@ -310,6 +310,61 @@ N|N
 	}
 }
 
+// TestLoadCSVSplit ensures that a split cannot happen in the middle of a row.
+func TestLoadCSVSplit(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	tmp, tmpCleanup := testutils.TempDir(t)
+	defer tmpCleanup()
+	ctx := context.Background()
+
+	const (
+		tableName   = "t"
+		csvName     = tableName + ".dat"
+		tableCreate = `
+			CREATE TABLE ` + tableName + ` (
+				i int primary key,
+				s string,
+				b int,
+				c int,
+				index (s),
+				index (i, s),
+				family (i, b),
+				family (s, c)
+			)
+		`
+		tableCSV = `5,STRING,7,9`
+	)
+
+	tablePath := filepath.Join(tmp, tableName)
+	dataPath := filepath.Join(tmp, csvName)
+
+	if err := ioutil.WriteFile(tablePath, []byte(tableCreate), 0666); err != nil {
+		t.Fatal(err)
+	}
+	if err := ioutil.WriteFile(dataPath, []byte(tableCSV), 0666); err != nil {
+		t.Fatal(err)
+	}
+	// sstMaxSize = 1 should put each index (could be more than one KV due to
+	// column families) in its own SST.
+	csv, kv, sst, err := sqlccl.LoadCSV(ctx, tablePath, []string{dataPath}, tmp, 0 /* comma */, 0 /* comment */, nil /* nullif */, 1 /* sstMaxSize */, tmp)
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Only a single input row.
+	if csv != 1 {
+		t.Fatalf("read %d rows, expected %d", csv, 1)
+	}
+	// Should produce 4 kvs: 2 families on PK + 2 indexes.
+	if kv != 4 {
+		t.Fatalf("created %d KVs, expected %d", kv, 4)
+	}
+	// But only 3 SSTs because the first 2 KVs should be in same SST.
+	if sst != 3 {
+		t.Fatalf("created %d SSTs, expected %d", sst, 3)
+	}
+}
+
 // TODO(dt): switch to a helper in sampledataccl.
 func makeCSVData(
 	t testing.TB, in string, numFiles, rowsPerFile int,

